### PR TITLE
Feature/camera-controller

### PR DESCRIPTION
**Pull request by PiterGroot**
All cameras are now cinemachine cameras.

**How to test:**
Go to develop scene and play normally. When you arrive at a station, switch to neutral gear. This will activate the "Enter" station camera transition. Increase your train gear to activate the "Exit" station camera transition.

You can obviously test this on any station. Train camera will now also shake a bit when going at its max speed (very subtle, feedback welcome) 

stationCam

To test the local player tracker, enter the game and play normally. Make sure to parent the players to one of the wagons manually in the hierarchy (since the parenting features is still being made)

### CODE REVIEW

File: Assets/Prefabs/Camera/Dolly Track.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Dolly Track.prefab:1]: Remove the space from the file name.

File: Assets/Prefabs/Camera/Station Camera.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Station Camera.prefab:1]: See previous comment.

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+using Cinemachine;
+using UnityEngine;
+
+using DerailedDeliveries.Framework.Utils;
+using System;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]:
```suggestion
using Cinemachine;
using UnityEngine;
using System;

using DerailedDeliveries.Framework.Utils;
```

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    /// <summary>
+    /// Class responsible for controlling and switching between virtual cameras.
+    /// </summary>
+    public class CameraManager : AbstractSingleton<CameraManager>
+    {
+        [SerializeField]
+        private CinemachineVirtualCamera[] _stationCameras;
+
+        [field: SerializeField]
+        public CinemachineVirtualCamera TrainCamera { get; private set; }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]: Missing summary.

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+                if (dSqrToTarget < closestDistanceSqr)
+                {
+                    if (!CheckIgnore(ignore, virtualCamera))
+                        continue;
+
+                    closestDistanceSqr = dSqrToTarget;
+                    bestTarget = virtualCamera;
+                }
+            }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]: Perhaps the use of a regular for loop would be better?

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+            distance = Mathf.Sqrt(closestDistanceSqr);
+            return Array.IndexOf(_stationCameras, bestTarget);
+        }
+
+        /// <summary>
+        /// Method to change priority to a specific camera, disables all other cameras.
+        /// </summary>
+        /// <param name="targetCamera"></param>
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]: Empty param for summary.

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            return true;
+        }
+
+        /// <summary>
+        /// Helper method responsible for getting the nearest virtual camera based on an origin position.
+        /// </summary>
+        /// <param name="originPosition">Position from which to look from.</param>
+        /// <param name="distance">Distance to the nearest camera.</param>
+        /// <param name="ignore"></param>
+        /// <returns></returns>
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]: Empty param ignore and empty return in summary.

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
+using UnityEngine.InputSystem;
+using Cinemachine;
+using UnityEngine;
+
+using DerailedDeliveries.Framework.Utils;
+using DerailedDeliveries.Framework.Camera;
+using FishNet.Object;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs]:
```suggestion
using UnityEngine.InputSystem;
using FishNet.Object;
using Cinemachine;
using UnityEngine;

using DerailedDeliveries.Framework.Utils;
using DerailedDeliveries.Framework.Camera;
```

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        [SerializeField]
+        private Vector3[] _stationPositions;
+
+        private float _distance;
+        private TrainController _trainController;
+       
+        private void Awake()
+        {
+            _trainController = GetComponent<TrainController>();
+        }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs]:
```suggestion
        private void Awake() => _trainController = GetComponent<TrainController>();
```

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+            anim.SetTrigger("Enter");
+        }
+       
+        [Server]
+        private bool CanParkTrain()
+        {
+            return TrainEngine.Instance.CurrentSpeedIndex == 0 
+                && Mathf.Abs(TrainEngine.Instance.CurrentSpeed) < .1f;
+        }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs]:
```suggestion
        private bool CanParkTrain() =>
            TrainEngine.Instance.CurrentSpeedIndex == 0 && Mathf.Abs(TrainEngine.Instance.CurrentSpeed) < .1f;
```

File: Assets/Scripts/Framework/Train/TrainController.cs [hunk 9]
```diff
@@ -2,6 +2,8 @@
 using FishNet.Object;
 using UnityEngine;
 using System;
+using DerailedDeliveries.Framework.Camera;
+using Cinemachine;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/TrainController.cs]:
```suggestion
using FishNet.Object;
using Cinemachine;
using UnityEngine;
using System;

using DerailedDeliveries.Framework.Camera;
```

File: Assets/Scripts/Framework/Train/TrainController.cs [hunk 10]
```diff
@@ -118,15 +123,20 @@ private void OnTick()
         [ObserversRpc(RunLocally = true)]
         private void MoveTrain(float distanceAlongSpline)
         {
+            Vector3 positionSum = Vector3.zero;
             UpdateWagonPosition(_frontWagon, distanceAlongSpline);
 
+            positionSum += _frontWagon.position;
+
             int wagons = _followingWagons.Length;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/TrainController.cs]:
```suggestion
            int wagonsAmount = _followingWagons.Length;
```
I name these in a way that you can instantly link them to the for loop. This is just a suggestion and not a must.

File: Assets/Scripts/Framework/Train/TrainEngine.cs [hunk 11]
```diff
@@ -92,14 +99,20 @@ public TrainEngineState EngineState
...
         private float _startFriction;
+        private float _startCameraNoiseAmplitude;
         
         private bool _isBraking;
 
-        private void Awake() => _trainController = GetComponent<TrainController>();
+        private void Awake()
+        {
+            _trainController = GetComponent<TrainController>();
+        }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/TrainEngine.cs]:
```suggestion
        private void Awake() => _trainController = GetComponent<TrainController>();
```

**Review by Joost-Robat — COMMENTED** [review]
Looks good. Minor feedback.

File: Assets/Scripts/Framework/Camera/CameraManager.cs (hunk 3 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/Camera/CameraManager.cs]: resolved

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                float dSqrToTarget = directionToTarget.sqrMagnitude;
+
+                if (dSqrToTarget < closestDistanceSqr)
+                {
+                    if (!CheckIgnore(ignore, virtualCamera))
+                        continue;
+
+                    closestDistanceSqr = dSqrToTarget;
+                    bestTarget = virtualCamera;
+                }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]:
```suggestion
                if (dSqrToTarget > closestDistanceSqr || !CheckIgnore(ignore, virtualCamera))
                    continue;
                
                closestDistanceSqr = dSqrToTarget;
                bestTarget = virtualCamera;
```

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        /// <summary>
+        /// Method for setting a specific camera to active using priority, disables all other cameras.
+        /// </summary>
+        /// <param name="targetCamera">Camera to set active.</param>
+        public void ChangeActiveCamera(CinemachineVirtualCamera targetCamera)
+        {
+            int cameras = StationCameras.Length;
+            for (int i = 0; i < cameras; i++)
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs]:
```suggestion
            int camerasAmount = StationCameras.Length;
            for (int i = 0; i < camerasAmount; i++)
```

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+            targetCamera.Priority = 1;
+        }
+
+        private bool CheckIgnore(CinemachineVirtualCamera[] ignore, CinemachineVirtualCamera cameraToCheck)
+        {
+            for (int i = 0; i < ignore.Length; i++)
+            {
+                if (ignore[i] == cameraToCheck)
+                    return false;
+            }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Camera/CameraManager.cs:77]:
```suggestion
            for (int i = 0; i < ignore.Length; i++; ignore[i] == cameraToCheck)
                return false;
```

**Review by Joost-Robat — CHANGES_REQUESTED** [review]
Looks good, only minor chages are required.

File: Assets/Prefabs/Camera/Virtual Camera.prefab.meta
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Virtual Camera.prefab.meta:1]: Please remove the space from the naming of this file.

File: Assets/Prefabs/Camera/Virtual Camera.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Virtual Camera.prefab:1]: Please remove the space in the naming of this file.

File: Assets/Prefabs/Camera/Target Group.prefab.meta
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Target Group.prefab.meta:1]: Please remove the space in the naming of this file.

File: Assets/Prefabs/Camera/Target Group.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Camera/Target Group.prefab:1]: Please remove the space in the naming of this file.

**Review by Joost-Robat — CHANGES_REQUESTED** [review]
Only some name changes are required and then it's ready for merge. Nice job!

File: Assets/Animations.meta
Comment (lucasvdiepen) [inline Assets/Animations.meta:1]: Move to `Assets/Art/3D/Animations`.

File: Assets/Animations/Station/Station.controller
Comment (lucasvdiepen) [inline Assets/Animations/Station/Station.controller:1]: Change name to `AC_Station`.

File: Assets/Animations/Station/StationEnter.anim
Comment (lucasvdiepen) [inline Assets/Animations/Station/StationEnter.anim:1]: Change name to `A_StationEnter`.

File: Assets/Animations/Station/StationLeave.anim
Comment (lucasvdiepen) [inline Assets/Animations/Station/StationLeave.anim:1]: Change name to `A_StationLeave`.

File: Assets/Art/2D/Train/VH_TrainBaseFrontCabin.fbx
Comment (lucasvdiepen) [inline Assets/Art/2D/Train/VH_TrainBaseFrontCabin.fbx:1]: Why is this model added in the `2D` folder?

File: Assets/Art/2D/Train/VH_TrainBaseSideCabin.fbx
Comment (lucasvdiepen) [inline Assets/Art/2D/Train/VH_TrainBaseSideCabin.fbx:1]: Same for this model.

File: Assets/Art/2D/Train/VH_TrainFrontCabin.fbx
Comment (lucasvdiepen) [inline Assets/Art/2D/Train/VH_TrainFrontCabin.fbx:1]: Same for this model. And also a duplicate of `VH_TrainBaseFrontCabin.fbx`.

File: Assets/Art/2D/Train/VH_TrainSideCabin.fbx
Comment (lucasvdiepen) [inline Assets/Art/2D/Train/VH_TrainSideCabin.fbx:1]: Same for this model.

File: Assets/Art/Materials/Train/M_TrainBase.mat.meta [hunk 15]
```diff
@@ -2,7 +2,7 @@ fileFormatVersion: 2
 guid: dc5c9228edc34ad4e8a16c86b2a5d59d
 NativeFormatImporter:
   externalObjects: {}
-  mainObjectFileID: 2100000
+  mainObjectFileID: 0
```
Comment (lucasvdiepen) [inline Assets/Art/Materials/Train/M_TrainBase.mat.meta:5]: What does this change do?

**Comment by PiterGroot** [comment]
TODO: Tweak parking values

File: Assets/Scripts/Framework/Camera/CameraManager.cs [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        /// <returns>Index of nearest camera.</returns>
+        public int GetNearestCamera(Vector3 originPosition, out float distance, params CinemachineVirtualCamera[] ignore)
+        {
+            CinemachineVirtualCamera bestTarget = null;
+            float closestDistanceSqr = Mathf.Infinity;
+
+            foreach (CinemachineVirtualCamera virtualCamera in StationCameras)
+            {
+                Vector3 directionToTarget = virtualCamera.transform.position - originPosition;
+                float dSqrToTarget = directionToTarget.sqrMagnitude;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Camera/CameraManager.cs:41]: Maybe change the `dSqrToTarget` variable name. It is not very clear to me.

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 17]
```diff
@@ -0,0 +1,101 @@
+using FishNet.Object;
+using Cinemachine;
+using UnityEngine;
+
+using DerailedDeliveries.Framework.Camera;
+using DerailedDeliveries.Framework.Utils;
+
+namespace DerailedDeliveries.Framework.Train
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs:8]: Wrong namespace.

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        private void OnEnable()
+            => TrainEngine.Instance.OnSpeedChanged += HandleSpeedChanged;
+
+        private void OnDisable()
+            => TrainEngine.Instance.OnSpeedChanged -= HandleSpeedChanged;
+
+        private void HandleSpeedChanged(float newSpeed)
+        {
+            if (newSpeed <= _minTrainSpeedToPark && !IsParked)
+                TryParkTrainAtClosestStation();
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs:54]: `TryParkTrainAtClosestStation()` should only be called on the server. Currently all client will send a rpc to the server.

File: Assets/Scripts/Framework/Camera/TrainStationController.cs [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+            if (_distance > _minRangeToNearestStation)
+                return;
+
+            TryParkTrain(nearestCameraIndex);
+        }
+
+        [ObserversRpc(RunLocally = true, BufferLast = true)]
+        private void TryParkTrain(int nearestStationCameraIndex)
+        {
+            TrainEngine.Instance.SetEngineState(TrainEngineState.Inactive);
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Camera/TrainStationController.cs:91]: I don't think we need to disable the engine when we park.

File: Assets/Scripts/Framework/Gameplay/Interactions/CoalGrabbable.cs
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Interactions/CoalGrabbable.cs:1]: Why is this file moved?

File: Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/UseableGrabbable.cs [hunk 20]
```diff
@@ -23,9 +23,6 @@ private protected override void UseGrabbable(Interactor interactor)
             if(targetInteractable != null && RunInteract(targetInteractable))
                 return;
 
-            if(IsDeinitializing)
-                return;
-
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/UseableGrabbable.cs:28]: Should not be removed.

File: Assets/Scripts/Framework/PlayerManagement/PlayerManager.cs [hunk 21]
```diff
@@ -116,6 +120,7 @@ public void PlayerJoined(PlayerId playerId)
             }
 
             OnPlayerJoined?.Invoke(playerId);
+            _targetGroup.AddMember(playerId.transform, 1, 1);
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/PlayerManagement/PlayerManager.cs:123]: Listen to the `OnPlayerJoined` event in another script instead of declaring Cinemachine code here.

File: Assets/Scripts/Framework/PopupManegement/Popup.cs [hunk 22]
```diff
@@ -21,11 +21,6 @@ public class Popup : MonoBehaviour
         [SerializeField]
         private float _hoverDuration = 1.5f;
 
-        /// <summary>
-        /// Gets whether the popup is showing or not.
-        /// </summary>
-        public bool IsShowing { get; private set; }
-
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/PopupManegement/Popup.cs:28]: Should not be removed.

File: Assets/Scripts/Framework/PopupManegement/PopupTriggerActivator.cs
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/PopupManegement/PopupTriggerActivator.cs:1]: This file should also not be changed.

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

File: Assets/Animations.meta
Comment (PiterGroot) [reply Assets/Animations.meta:1]: Fixed in new branch.

File: Assets/Animations/Station/Station.controller
Comment (PiterGroot) [reply Assets/Animations/Station/Station.controller:1]: Fixed in new branch.

File: Assets/Animations/Station/StationEnter.anim
Comment (PiterGroot) [reply Assets/Animations/Station/StationEnter.anim:1]: Fixed in new branch.

File: Assets/Animations/Station/StationLeave.anim
Comment (PiterGroot) [reply Assets/Animations/Station/StationLeave.anim:1]: Fixed in new branch.

File: Assets/Art/2D/Train/VH_TrainBaseFrontCabin.fbx
Comment (PiterGroot) [reply Assets/Art/2D/Train/VH_TrainBaseFrontCabin.fbx:1]: Fixed in new branch.

File: Assets/Art/2D/Train/VH_TrainBaseSideCabin.fbx
Comment (PiterGroot) [reply Assets/Art/2D/Train/VH_TrainBaseSideCabin.fbx:1]: Fixed in new branch.

File: Assets/Art/2D/Train/VH_TrainFrontCabin.fbx
Comment (PiterGroot) [reply Assets/Art/2D/Train/VH_TrainFrontCabin.fbx:1]: Fixed in new branch.

File: Assets/Art/2D/Train/VH_TrainSideCabin.fbx
Comment (PiterGroot) [reply Assets/Art/2D/Train/VH_TrainSideCabin.fbx:1]: Fixed in new branch.

File: Assets/Art/Materials/Train/M_TrainBase.mat.meta (hunk 15 above)
Comment (PiterGroot) [reply Assets/Art/Materials/Train/M_TrainBase.mat.meta:5]: Fixed and reverted in new branch.

File: Assets/Scripts/Framework/Camera/CameraManager.cs (hunk 16 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/Camera/CameraManager.cs:41]: Agreed, I changed it to `distanceSquared` in new branch.

File: Assets/Scripts/Framework/Camera/TrainStationController.cs (hunk 19 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/Camera/TrainStationController.cs:91]: Yes, it actually created some problems. Removed in new branch

File: Assets/Scripts/Framework/Gameplay/Interactions/CoalGrabbable.cs
Comment (PiterGroot) [reply Assets/Scripts/Framework/Gameplay/Interactions/CoalGrabbable.cs:1]: Fixed in new branch

File: Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/UseableGrabbable.cs (hunk 20 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/UseableGrabbable.cs:28]: Fixed in new branch

File: Assets/Scripts/Framework/PlayerManagement/PlayerManager.cs (hunk 21 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/PlayerManagement/PlayerManager.cs:123]: Fixed in new branch

File: Assets/Scripts/Framework/PopupManegement/Popup.cs (hunk 22 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/PopupManegement/Popup.cs:28]: Fixed in new branch

File: Assets/Scripts/Framework/PopupManegement/PopupTriggerActivator.cs
Comment (PiterGroot) [reply Assets/Scripts/Framework/PopupManegement/PopupTriggerActivator.cs:1]: Fixed in new branch

**Comment by PiterGroot** [comment]
Closed because this branch is a bit corrupted, fixed in new [branch](https://github.com/lucasvdiepen/DerailedDeliveries/pull/38).
